river/tree/base.py

```diff
@@ -28,14 +28,29 @@ def __init__(self, *children):
     def next(self, x) -> Union["Branch", "Leaf"]:
         """Move to the next node down the tree."""
 
-    def walk(self, x) -> Iterable[Union["Branch", "Leaf"]]:
+    @abc.abstractmethod
+    def most_common_path(self) -> Tuple[int, Union["Leaf", "Branch"]]:
```

Comment:
This shouldn't be here. The most common path logic should be specific to each, and should not be part of the `Tree` interface.

Reply:
I disagree :P

Every branch calculates the most common path differently. That's why I created the abstract method. This is the core functionality to deal with missing split features.

I think this is generic enough to belong to the base class. Do you have other suggestions?

Reply:
Ok very clear! How about just handling the "go to the most branch child" in the `next` function? It's what I've been doing for half-space trees

Reply:
The thing is, sometimes we need to do that "manually". Let me give you another example. Suppose two cases while learning:

1. We receive an instance where a split feature is missing
2. We receive an instance where a new category emerges in one of the split features

In the first case, the tree must pick the most traversed path and return the reached leaf. In the second one, the tree must stop the traversal in the internal node and create a new branch (and leaf) to accommodate the new category. Each tree has specific sets of leaves (that also depend on the prediction strategy selected by the user), that require specific sets of parameters to be constructed. That's why they are usually built by the "main tree" rather than by branches.

Therefore, I decided to expose the `most_common_path`, so that the trees could wisely use it. In summary, it's not only a matter of taking the "main road", but also deciding when to travel :D

Reply:
Mmm I see! But you could also argue that the leaf create could happen in the `next`, and the branch could just have a point to the main tree, no? I'm just trying to weigh all possibilities.

Reply:
It could do the job, perhaps. I need to think a little bit about the weirdest trees though. Anyhow, it sounds to me that we will end up delegating too much power to a single method, haha. Besides, all new tree implementations would need to account for this and possible other corner cases in `next`. And yet, `next` would need to know when to take one type of action or another.

Currently, `next` tries to obtain the next node in the path, if it's not possible an error is thrown. Then `walk` decides what to do, depending on `until_leaf` (by the way, any ideas for a better name are welcome). The walk can either stop at the branch or continue by following the most traversed path.

I also do not appreciate very much having to implement `most_common_path` every time. But it still sounds like a good compromise. (Kind of) quoting Jacob: otherwise, we might end up complicating even more the maze that is the trees.